File: 백준/Gold/2098. 외판원 순회/외판원 순회.py
import sys
input = sys.stdin.readline

N = int(input())
INF = int(1e9)


graph = [list(map(int, input().split())) for _ in range(N)]
dp = [[-1] * (1 << N) for _ in range(N)]

def tsp(x, bitState):
    if (bitState == (1 << N) - 1): # 모든 도시를 방문 했다면
        if (graph[x][0] != 0): # x에서 출발점으로 갈 수 있다면
            return graph[x][0]
        else: # 출발점으로 갈 수 없을 경우
            return INF

    if dp[x][bitState] != -1: # 이미 최소비용이 계산되어 있을 경우
        return dp[x][bitState]

    dp[x][bitState] = INF
    for i in range(N):
        if graph[x][i] == 0 or (bitState & (1 << i) > 0): # 경로가 있고 방문하지 않은 경우
            continue
        dp[x][bitState] = min(dp[x][bitState], tsp(i, bitState | (1 << i)) + graph[x][i])

    return dp[x][bitState]

# 시작 지점은 어느 집이 되어도 상관없음
# 0 → 1 → 2 → 3 → 0랑 2 → 3 → 0 → 1 → 2처럼 순회하는 경로의 정답은 동일함.
print(tsp(0, 1))

File: 백준/Gold/2098. 외판원 순회/외판원 순회.java
import java.io.*;
import java.util.StringTokenizer;
import java.util.Arrays;
public class Main {
    static int N;
    static int[][] map;
    static int[][] dp;
    static StringTokenizer st = null;
    static int INF = 17000000;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        map = new int[N][N];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < N; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }
        // dp[i][j] : 현재 있는 도시가 i이고 이미 방문한 도시들의 집합이 j일때 방문하지 않은 나머지 도시들을 모두 방문한 뒤 출발도시로로 돌아올때 드는 최소 비용
        dp = new int[N][1<<N];
        for (int i = 0; i < N; i++) {
            Arrays.fill(dp[i], -1);
        }
        System.out.println(tsp(0, 1)); // 0번 도시부터 탐색, bitState : 0001
    }

    static int tsp(int x, int bitState) {
        // 모든 도시 방문 완료
        if (bitState == (1 << N) - 1) {
            if (map[x][0] == 0) { // 경로 없으면 탐색 무효화
                return INF;
            } else {
                return map[x][0];
            }
        }

        if (dp[x][bitState] != -1) { // 이미 방문한 도시
            return dp[x][bitState];
        }

        // x 도시 방문 표시
        dp[x][bitState] = INF;

        for (int i = 0; i < N; i++) {
            // next : 다음으로 방문할 i 도시
            int next = bitState | (1 << i);

            // 경로가 있고, 방문 하지 않은 경우
            if (map[x][i] != 0 && (bitState & (1 << i)) == 0 ) {
                dp[x][bitState] = Math.min(dp[x][bitState], tsp(i, next) + map[x][i]);
            }
        }

        return dp[x][bitState];
    }
}
